### PR TITLE
Initial implementation of DuckDb genotype storage

### PR DESCRIPTION
**Pull request by lchorbadjiev**
## Background

DuckDb is an embeddable OLAP database. It allows the implementation of genotype storage that works locally without relying on an OLAP database cluster.

## Aim

Provide an initial implementation of DuckDb genotype storage.

## Implementation

The DuckDb storage uses `Schema2ImportStorage` and `SqlSchema2Variants`. We had to provide `DuckDbQueryDialect` and to make some modification in how SQL `Dialect` classes are used.

### CODE REVIEW

File: dae/dae/duckdb_storage/duckdb_genotype_storage.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+from dae.parquet.partition_descriptor import PartitionDescriptor
+from dae.genotype_storage.genotype_storage import GenotypeStorage
+from dae.schema2_storage.schema2_import_storage import Schema2DatasetLayout
+from dae.duckdb_storage.duckdb_variants import DuckDbVariants
+
+logger = logging.getLogger(__name__)
+
+
+class DuckDbGenotypeStorage(GenotypeStorage):
+    """Defines Google Cloud Platform (GCP) genotype storage."""
```
Comment (iordanivanov) [inline dae/dae/duckdb_storage/duckdb_genotype_storage.py:19]: This docstring should probably be updated

File: dae/dae/duckdb_storage/duckdb_genotype_storage.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        return "duckdb"
+
+    def start(self):
+        if self.connection:
+            logger.warning(
+                "starting already started DuckDb genotype storage: <%s>",
+                self.storage_id)
+            return self
+            # raise ValueError(
+            #     f"already started DuckDb storage <{self.storage_id}>")
```
Comment (iordanivanov) [inline dae/dae/duckdb_storage/duckdb_genotype_storage.py:62]: Commented-out code

File: dae/dae/duckdb_storage/duckdb_variants.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    ):
+        super().__init__(
+            DuckDbQueryDialect(),
+            db,
+            family_variant_table,
+            summary_allele_table,
+            pedigree_table,
+            meta_table,
+            gene_models)
+        # assert db, db
```
Comment (iordanivanov) [inline dae/dae/duckdb_storage/duckdb_variants.py:172]: Commented-out code

File: dae/dae/duckdb_storage/duckdb_variants.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                schema_content = row[0]
+        return dict(line.split("|") for line in schema_content.split("\n"))
+
+    def _fetch_schema(self, table) -> dict[str, str]:
+        return {}
+
+    def _fetch_pedigree(self):
+        query = f"SELECT * FROM {self.pedigree_table}"
+
+        # ped_df = pandas_gbq.read_gbq(q, project_id=self.gcp_project_id)
```
Comment (iordanivanov) [inline dae/dae/duckdb_storage/duckdb_variants.py:220]: Commented-out code

File: dae/dae/query_variants/sql/schema2/base_variants.py [hunk 5]
```diff
@@ -32,7 +32,7 @@ def __init__(
             pedigree_table,
             meta_table,
             gene_models=None):
-        assert db
+        # assert db
```
Comment (iordanivanov) [inline dae/dae/query_variants/sql/schema2/base_variants.py:35]: Commented-out code

File: dae/dae/query_variants/sql/schema2/family_builder.py [hunk 6]
```diff
@@ -64,10 +64,12 @@ def _query_columns(self):
...
-            inner = "fa.allele_in_members"
-            if self.dialect.add_unnest_in_join():
-                inner = f"UNNEST({inner})"
-            self._add_to_product(f"\n    JOIN\n    {inner} AS pi")
+            self._add_to_product(
+                self.dialect.build_array_join("fa.allele_in_members", "pi"))
+            # inner = "fa.allele_in_members"
+            # if self.dialect.add_unnest_in_join():
+            #     inner = f"UNNEST({inner})"
+            # self._add_to_product(f"\n    JOIN\n    {inner} AS pi")
```
Comment (iordanivanov) [inline dae/dae/query_variants/sql/schema2/family_builder.py:72]: Commented-out code

File: dae/dae/query_variants/sql/schema2/family_builder.py [hunk 7]
```diff
@@ -78,12 +80,15 @@ def _build_join(self, genes=None, effect_types=None):
...
+            self._add_to_product(
+                self.dialect.build_array_join("sa.effect_gene", "eg")
             )
-            self._add_to_product(f"\n    JOIN\n    {inner_clause} AS eg")
+            # inner_clause = (
+            #     "UNNEST(sa.effect_gene)"
+            #     if self.dialect.add_unnest_in_join()
+            #     else "sa.effect_gene"
+            # )
+            # self._add_to_product(f"\n    JOIN\n    {inner_clause} AS eg")
```
Comment (iordanivanov) [inline dae/dae/query_variants/sql/schema2/family_builder.py:91]: Commented-out code

File: dae/dae/query_variants/sql/schema2/summary_builder.py [hunk 8]
```diff
@@ -55,12 +55,14 @@ def _build_from(self):
...
-            )
-            self._add_to_product(f"\n    JOIN\n    {inner_clause} AS eg")
+            self._add_to_product(
+                self.dialect.build_array_join("sa.effect_gene", "eg"))
+            # inner_clause = (
+            #     "UNNEST(sa.effect_gene)"
+            #     if self.dialect.add_unnest_in_join()
+            #     else "sa.effect_gene"
+            # )
+            # self._add_to_product(f"\n    JOIN\n    {inner_clause} AS eg")
```
Comment (iordanivanov) [inline dae/dae/query_variants/sql/schema2/summary_builder.py:65]: Commented-out code

File: dae/dae/utils/fs_utils.py [hunk 9]
```diff
@@ -85,6 +87,14 @@ def sign(filename: str) -> str:
         return filename
 
 
+def copy(dest: str, src: str) -> None:
+    shutil.copytree(src, dest, dirs_exist_ok=True)
+    # src_mapper = fsspec.get_mapper(src)
+    # dest_mapper = fsspec.get_mapper(dest)
+    # for k in src_mapper:
+    #     dest_mapper[k] = src_mapper[k]
```
Comment (iordanivanov) [inline dae/dae/utils/fs_utils.py:95]: Commented-out code

File: dae/dae/utils/fs_utils.py [hunk 10]
```diff
@@ -1,10 +1,12 @@
 import datetime
 import os
+import shutil
 from pathlib import Path
 from urllib.parse import urlparse
 from typing import Optional, Union, cast
 
 from fsspec.core import url_to_fs
+# import fsspec
```
Comment (iordanivanov) [inline dae/dae/utils/fs_utils.py:9]: Commented-out import

**Review by iordanivanov — APPROVED** [review]

File: dae/dae/duckdb_storage/duckdb_genotype_storage.py (hunk 1 above)
Comment (lchorbadjiev) [reply dae/dae/duckdb_storage/duckdb_genotype_storage.py:19]: fixed

File: dae/dae/duckdb_storage/duckdb_genotype_storage.py (hunk 2 above)
Comment (lchorbadjiev) [reply dae/dae/duckdb_storage/duckdb_genotype_storage.py:62]: fixed

File: dae/dae/duckdb_storage/duckdb_variants.py (hunk 3 above)
Comment (lchorbadjiev) [reply dae/dae/duckdb_storage/duckdb_variants.py:172]: fixed

File: dae/dae/duckdb_storage/duckdb_variants.py (hunk 4 above)
Comment (lchorbadjiev) [reply dae/dae/duckdb_storage/duckdb_variants.py:220]: fixed

File: dae/dae/query_variants/sql/schema2/base_variants.py (hunk 5 above)
Comment (lchorbadjiev) [reply dae/dae/query_variants/sql/schema2/base_variants.py:35]: fixed

File: dae/dae/query_variants/sql/schema2/family_builder.py (hunk 6 above)
Comment (lchorbadjiev) [reply dae/dae/query_variants/sql/schema2/family_builder.py:72]: fixed

File: dae/dae/query_variants/sql/schema2/family_builder.py (hunk 7 above)
Comment (lchorbadjiev) [reply dae/dae/query_variants/sql/schema2/family_builder.py:91]: fixed

File: dae/dae/query_variants/sql/schema2/summary_builder.py (hunk 8 above)
Comment (lchorbadjiev) [reply dae/dae/query_variants/sql/schema2/summary_builder.py:65]: fixed

File: dae/dae/utils/fs_utils.py (hunk 10 above)
Comment (lchorbadjiev) [reply dae/dae/utils/fs_utils.py:9]: fixed

**Pull request merged**